src/v/cloud_storage/segment_meta_cstore.cc

```diff
@@ -563,65 +573,135 @@ class segment_meta_cstore::impl
+                return write_buffer.crbegin()->second;
+            }
+            // fallthrough
+        }
         return _col.last_segment();
     }
 
-    void insert(const segment_meta& m) { _col.append(m); }
+    void insert(const segment_meta& m) {
+        // TODO optimize to bypass this write?
+        write_buffer[m.base_offset] = m;
+        if (write_buffer.size() > max_buffer_entries) {
+            flush_write_buffer();
+        }
+    }
 
-    auto inflated_actual_size() const { return _col.inflated_actual_size(); }
+    auto inflated_actual_size() const {
+        // TODO how to deal with write_buffer?
+        return _col.inflated_actual_size();
+    }
 
-    size_t size() const { return _col.size(); }
+    size_t size() const {
+        // TODO how to deal with write_buffer?
+        return _col.size();
+    }
 
-    bool empty() const { return _col.empty(); }
+    bool empty() const { return write_buffer.empty() && _col.empty(); }
 
-    bool contains(model::offset o) { return _col.contains(o()); }
+    bool contains(model::offset o) {
+        return write_buffer.contains(o) || _col.contains(o());
+    }
 
     void prefix_truncate(model::offset new_start_offset) {
+        // TODO check this
+        auto new_begin = write_buffer.lower_bound(new_start_offset);
+        write_buffer.erase(write_buffer.begin(), new_begin);
         _col.prefix_truncate(new_start_offset());
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl>
     at_index(size_t ix) const {
+        // TODO deal with write_buffer
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.at_index(ix));
     }
 
-    auto serde_fields() { return std::tie(_col); }
+    auto serde_fields() {
+        if (!write_buffer.empty()) {
+            // TODO move check to flush_write_buffer if other methods use it
+            // write_buffer contains temp data that will occupy less data when
+            // compressed, so flush it
+            flush_write_buffer();
+        }
+        return std::tie(_col);
+    }
 
 private:
-    column_store _col;
+    void flush_write_buffer() {
+        auto last_committed_offset = _col.last_committed_offset().value_or(
+          model::offset::min());
+        // find the item separating "Replace" entries (those with a base offset
+        // lower or equal than last_committed_offset) and "Append" entries (all
+        // the other with a base offset greater than last_committed_offset, that
+        // can be simply appended)
+        auto partition_point = write_buffer.upper_bound(last_committed_offset);
+
+        auto replace_entries_begin = write_buffer.begin();
+        auto replace_entries_end = partition_point;
+        _col.replace_entries(replace_entries_begin, replace_entries_end);
+
+        auto append_entries_begin = partition_point;
+        auto append_entries_end = write_buffer.end();
+        // TODO bad_alloc safety?
```

Comment:
The `append` itself is transactional but here I think you need to make transaction span multiple elements, so modified version of the `append` is needed.